lib/datafile.js

```diff
@@ -403,32 +399,28 @@ _.extend(Reader.prototype, {
                 this.decoder
         );
 
-        if (block instanceof AvroErrors.BlockDelayReadError) {
-            cb(block);
-        } else {
-            // If the sync marker doesn't match, maybe there is no sync
-            // marker, try skipping back
-            if (block.sync && block.sync.toString() !== this.header.sync.toString()) {
-                self._fileBlock.skip(-SYNC_SIZE);
-            }
+        // If the sync marker doesn't match, maybe there is no sync
+        // marker, try skipping back
+        if (block.sync && block.sync.toString() !== this.header.sync.toString()) {
+            self._fileBlock.skip(-SYNC_SIZE);
+        }
 
-            var codec = this.header.meta['avro.codec'].toString();
-            this.decompressData(block.objects, codec, function(err, data) {
-                if (err) {
-                    self.emit('error',  err);
-                } else {
-                    if (data) {
-                        self._datumBlock.write(data);
-                        self.decoder.input(self._datumBlock);
-                        while (block.objectCount--) {
-                            var decoded = self.datumReader.read(self.decoder);
-                            if (decoded !== null) self.push(decoded);
-                        }
-                        cb();
+        var codec = this.header.meta['avro.codec'].toString();
+        this.decompressData(block.objects, codec, function(err, data) {
+            if (err) {
+                self.emit('error',  err);
+            } else {
+                if (data) {
+                    self._datumBlock.write(data);
+                    self.decoder.input(self._datumBlock);
+                    while (block.objectCount--) {
+                        var decoded = self.datumReader.read(self.decoder);
+                        if (decoded !== null) self.push(decoded);
                     }
+                    cb();
                 }
-            });
-        }
+            }
+        });
     },
 
     _readBlocks: function(cb) {
```

Comment:
So, when you catch the `BlockDelayReadError` inside `_readBlocks`, you update the offset and then callback, but the callback seems to break the recursion that would otherwise happen without the error - does that mean `_readBlocks` is called again by a higher-up function? or maybe you need call `self._readBlocks(cb)` instead in the `catch` block (and maybe even queue the callback to happen after the next tick to be sure):
```javascript
try {
  self._readBlock(...);
} catch (e) {
  if (e instanceof AvroErrors.BlockDelayReadError) {
    self._readBlocks(cb); // possibly wrap in `process.nextTick` to allow it to wait for data
  } else throw e;
}
```